File: app/src/main/java/com/mohsenoid/rickandmorty/view/character/list/CharacterListPresenter.java
package com.mohsenoid.rickandmorty.view.character.list;

import com.mohsenoid.rickandmorty.data.DataCallback;
import com.mohsenoid.rickandmorty.data.exception.NoOfflineDataException;
import com.mohsenoid.rickandmorty.domain.Repository;
import com.mohsenoid.rickandmorty.model.CharacterModel;
import com.mohsenoid.rickandmorty.util.config.ConfigProvider;

import java.util.List;

public class CharacterListPresenter implements CharacterListContract.Presenter {

    private final Repository repository;
    private final ConfigProvider configProvider;

    private CharacterListContract.View view = null;
    private List<Integer> characterIds;

    public CharacterListPresenter(Repository repository, ConfigProvider configProvider) {
        this.repository = repository;
        this.configProvider = configProvider;
    }

    @Override
    public void bind(CharacterListContract.View view) {
        this.view = view;
    }

    @Override
    public void unbind() {
        view = null;
    }

    @Override
    public void setCharacterIds(List<Integer> characterIds) {
        this.characterIds = characterIds;
    }

    @Override
    public void loadCharacters() {
        if (view != null) view.showLoading();

        queryCharacters();
    }

    private void queryCharacters() {
        if (!configProvider.isOnline()) {
            if (view != null) view.showOfflineMessage(false);
        }

        repository.queryCharacters(characterIds, new DataCallback<List<CharacterModel>>() {

            @Override
            public void onSuccess(List<CharacterModel> characters) {
                if (view != null) {
                    view.setCharacters(characters);
                    view.hideLoading();
                }
            }

            @Override
            public void onError(Exception exception) {
                if (view != null) {
                    view.hideLoading();
                }

                if (exception instanceof NoOfflineDataException) {
                    if (view != null) view.onNoOfflineData();
                } else {
                    if (view != null) view.showMessage(exception.getMessage());
                }
            }
        });
    }

    @Override
    public void killCharacter(CharacterModel character) {
        if (!character.isAlive()) return;

        repository.killCharacter(character.getId(), new DataCallback<CharacterModel>() {

            @Override
            public void onSuccess(CharacterModel character) {
                view.updateCharacter(character);
            }

            @Override
            public void onError(Exception exception) {
                if (view != null) view.showMessage(exception.getMessage());
            }
        });
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/view/character/list/CharacterListPresenter.kt
package com.mohsenoid.rickandmorty.view.character.list

import com.mohsenoid.rickandmorty.domain.Repository
import com.mohsenoid.rickandmorty.domain.model.ModelCharacter
import com.mohsenoid.rickandmorty.domain.model.QueryResult
import com.mohsenoid.rickandmorty.util.config.ConfigProvider

class CharacterListPresenter(
    private val repository: Repository,
    private val configProvider: ConfigProvider
) : CharacterListContract.Presenter() {

    private var view: CharacterListContract.View? = null

    override fun bind(view: CharacterListContract.View) {
        this.view = view
    }

    override fun unbind() {
        view = null
    }

    override suspend fun loadCharacters(characterIds: List<Int>) {
        view?.showLoading()
        queryCharacters(characterIds)
    }

    private suspend fun queryCharacters(characterIds: List<Int>) {
        if (!configProvider.isOnline()) {
            view?.showOfflineMessage(isCritical = false)
        }

        when (val result = repository.getCharactersByIds(characterIds)) {
            is QueryResult.Successful -> view?.setCharacters(result.data)
            QueryResult.NoCache -> view?.onNoOfflineData()
        }

        view?.hideLoading()
    }

    override suspend fun killCharacter(character: ModelCharacter) {
        if (!character.isAlive) return

        when (val result = repository.killCharacter(character.id)) {
            is QueryResult.Successful -> view?.updateCharacter(result.data)
            QueryResult.NoCache -> view?.onNoOfflineData()
        }
    }
}
